### ISSUE TITLE
TF run produces incorrect diff and fails to apply

### ISSUE THREAD

**Issue by greatvovan** [grafana/terraform-provider-grafana#711]
### Terraform Version
* Terraform: 1.3.3
* Terraform Grafana Provider: 1.30.0
* Grafana: Cloud

### Affected Resource(s)
* grafana_rule_group

### Steps to Reproduce
For a clean setup, create a new folder (provision or manually).

1. Provision a group "default" with a single alert named "Foobar test". Use a minimal definition that works for your environment. Definition does not matter, it just has to succeed.
2. Add another (identical) minimalistic alert named "Foobar".
3. Create a TF run and look at the output.
4. Try to apply the run and observe the error.

### Expected Behavior
On steps `#3` and `#4` a new alert rule is added.

### Actual Behavior
Apply fails with the error:

```text
│ Error: status: 500, body: {"message":"failed to insert alert rules: a conflicting alert rule is found: rule title under the same organisation and folder should be unique","traceID":"19006e8578a22df0"}
│ 
│   with grafana_rule_group.default["group"],
│   on resources.rule_group.tf line 1, in resource "grafana_rule_group" "default":
│    1: resource "grafana_rule_group" "default" {
```

### More information
Apparently the issue comes from wrong matching of existing vs. provisioned alert rules. Instead of just adding a rule "Foobar", it tries to add "Foobar test" and rename "Foobar test" --> "Foobar". You can find confirmation of that in run output on step 3:

```
Terraform will perform the following actions:

  # grafana_rule_group.default["group"] will be updated in-place
  ~ resource "grafana_rule_group" "default" {
        id               = "bk_R1GDVz;default"
        name             = "default"
        # (3 unchanged attributes hidden)

      ~ rule {
          ~ name           = "Foobar test" -> "Foobar"
            # (7 unchanged attributes hidden)

            # (2 unchanged blocks hidden)
        }
      + rule {
          + condition      = "B"
          + exec_err_state = "Alerting"
          + for            = "0s"
          + labels         = {
              + "type" = "no_data_test"
            }
          + name           = "Foobar test"
          + no_data_state  = "Alerting"

          + data {
          # (hidden as irrelevant)
            }
        }
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```
Note renaming of the existing rule ^^^^^

### Possible solution
The problem is that the Terraform resource is group-centric, while provisioning API is rule-centric. From TF's perspective, it looks like changing a single grafana_rule_group item, but I can assume that under the hood the above gets converted into one `POST /api/v1/provisioning/alert-rules` and one `PUT /api/v1/provisioning/alert-rules/{UID}`, with the problem happening on the former.

Thus I see to alternatives:
1. Create grafana_alert_rule resource and manage matching of infrastructure vs TF using rule name. Grafana is maintaining unique names within a folder, so there should not be conflicts.
2. Fix conversion of TF run diff into API calls. Apparently rule name should be used again to achieve proper matching.

### Important Factoids
> Are there anything atypical about your accounts that we should know?

No, everything is standard and trivial.

### Example definitions
First run & apply:
```
resource "grafana_rule_group" "default" {
    name             = "default"
    folder_id        = "abcdefg"
    org_id           = 1
    interval_seconds = 60

    rule {
        condition      = "B"
        for            = "0s"
        labels         = { }
        name           = "Foobar test"

        data { ... }
    }
}
```

Second run:
```
resource "grafana_rule_group" "default" {
    name             = "default"
    folder_id        = "abcdefg"
    org_id           = 1
    interval_seconds = 60

    rule {
        condition      = "B"
        for            = "0s"
        labels         = { }
        name           = "Foobar test"

        data { ... }
    }
    rule {
        condition      = "B"
        for            = "0s"
        labels         = { }
        name           = "Foobar"

        data { ... }
    }
}
```

**Comment by svyatoslav-zakhryapin-stenn** [comment]
+1, just got the same issue

**Comment by greatvovan** [comment]
Any feedback from developers by chance? The issue is quite a deal breaker with terraforming alert rules.

The workaround seems to be putting every rule into a separate group. Not sure if there are any performance or other implications of this.

**Comment by alexweav** [comment]
Hmm, I was not able to reproduce this with the given Terraform definitions. I used the latest version of the provider (1.30.0), and a build from `main` of Grafana.

Here are the exact terraform definitions I used, which are nearly verbatim from the above with required fields fixed:
First apply:
```
resource "grafana_rule_group" "default" {
    name             = "default"
    folder_uid        = "abcdefg"
    org_id           = 1
    interval_seconds = 60

    rule {
        condition      = "B"
        for            = "0s"
        labels         = { }
        name           = "Foobar test"

        data {
          datasource_uid = "test"
          ref_id = "A"
          relative_time_range {
            from = 3600
            to = 0
          }
          model = "{}"
        }
    }
}
```

Second apply:
```
resource "grafana_rule_group" "default" {
    name             = "default"
    folder_uid        = "abcdefg"
    org_id           = 1
    interval_seconds = 60

    rule {
        condition      = "B"
        for            = "0s"
        labels         = { }
        name           = "Foobar test"

        data {
          datasource_uid = "test"
          ref_id = "A"
          relative_time_range {
            from = 3600
            to = 0
          }
          model = "{}"
        }
    }
    
    rule {
        condition      = "B"
        for            = "0s"
        labels         = { }
        name           = "Foobar"

        data {
          datasource_uid = "test"
          ref_id = "A"
          relative_time_range {
            from = 3600
            to = 0
          }
          model = "{}"
        }
    }
}
```

The provided diff was correct on the second apply too. It didn't show the first rule being replaced by the second rule, but rather only showed the second rule being added with some calculated fields in `model`.

This is also interesting, in that we have an integration test that appears to cover this exact use case, that's currently passing:
https://github.com/grafana/terraform-provider-grafana/blob/3690db2ce15559013842a5a83be8bae67780cbe1/grafana/resource_alerting_rule_group_test.go#L139C4-L150

I'm curious as to whether copy/pasting the above definitions also works for you? I wonder what else could be the cause? Was there something I missed?

> The workaround seems to be putting every rule into a separate group. Not sure if there are any performance or other implications of this.

This is perfectly fine to do - in fact, usually there are perf implications with putting many rules in the same group. Many groups is not a problem, in most cases it's actually preferred.

**Comment by greatvovan** [comment]
That's strange, on my end it has 100% reproducibility rate.

> I'm curious as to whether copy/pasting the above definitions also works for you?

You mean to test it verbatim, with `datasource_uid = "test"`? I do not have a source with such ID.

> Many groups is not a problem, in most cases it's actually preferred.

Cool, good to know that.

> we have an integration test that appears to cover this exact use case

What is the current method the provider uses to distinguish new resources from existing resources?

**Comment by svyatoslav-zakhryapin-stenn** [comment]
@alexweav could you try this as a third apply?
```
resource "grafana_rule_group" "default" {
    name             = "default"
    folder_uid        = "abcdefg"
    org_id           = 1
    interval_seconds = 60

    rule {
        condition      = "B"
        for            = "0s"
        labels         = { }
        name           = "Foobar test"

        data {
          datasource_uid = "test"
          ref_id = "A"
          relative_time_range {
            from = 3600
            to = 0
          }
          model = "{}"
        }
    }
    
    rule {
        condition      = "B"
        for            = "0s"
        labels         = { }
        name           = "Foobar baz"

        data {
          datasource_uid = "test"
          ref_id = "A"
          relative_time_range {
            from = 3600
            to = 0
          }
          model = "{}"
        }
    }

    rule {
        condition      = "B"
        for            = "0s"
        labels         = { }
        name           = "Foobar"

        data {
          datasource_uid = "test"
          ref_id = "A"
          relative_time_range {
            from = 3600
            to = 0
          }
          model = "{}"
        }
    }
}
```

This is what I get in the plan when trying to apply this:
```
  # grafana_rule_group.default will be updated in-place
  ~ resource "grafana_rule_group" "default" {
        id               = "abcdefg;default"
        name             = "default"
        # (3 unchanged attributes hidden)

      ~ rule {
            name           = "Foobar test"
            # (7 unchanged attributes hidden)

          ~ data {
              ~ model          = jsonencode(
                  ~ {
                      - intervalMs    = 1000 -> null
                      - maxDataPoints = 43200 -> null
                    }
                )
                # (2 unchanged attributes hidden)

                # (1 unchanged block hidden)
            }
        }
      ~ rule {
          ~ name           = "Foobar" -> "Foobar baz"
            # (7 unchanged attributes hidden)

          ~ data {
              ~ model          = jsonencode(
                  ~ {
                      - intervalMs    = 1000 -> null
                      - maxDataPoints = 43200 -> null
                    }
                )
                # (2 unchanged attributes hidden)

                # (1 unchanged block hidden)
            }
        }
      + rule {
          + condition      = "B"
          + exec_err_state = "Alerting"
          + for            = "0s"
          + name           = "Foobar"
          + no_data_state  = "NoData"

          + data {
              + datasource_uid = "test"
              + model          = jsonencode({})
              + ref_id         = "A"

              + relative_time_range {
                  + from = 3600
                  + to   = 0
                }
            }
        }
    }
```

As you can see, the provider tries to replace the "Foobar" rule.

**Comment by olekgo** [comment]
> I'm curious as to whether copy/pasting the above definitions also works for you? 

Worked for me too (had to change `folder_uid`)

> Here are the exact terraform definitions I used, which are nearly verbatim from the above with required fields fixed:

@alexweav With a small modification to the code you've linked, I was able to reproduce the issue (grafana version `9.3.1`, provider `1.30.0`). The gist is that I applied the rule `Foobar` first, and `Foobar test` second. Likely something to do with alphabetic order


First apply:

```
resource "grafana_rule_group" "default" {
    name             = "default"
    folder_uid        = "7WfjWOKVk"
    org_id           = 1
    interval_seconds = 60
 
    rule {
        condition      = "B"
        for            = "0s"
        labels         = { }
        name           = "Foobar"

        data {
          datasource_uid = "test"
          ref_id = "A"
          relative_time_range {
            from = 3600
            to = 0
          }
          model = "{}"
        }
    }
}

```

Second apply:
> 
```
resource "grafana_rule_group" "default" {
    name             = "default"
    folder_uid        = "7WfjWOKVk"
    org_id           = 1
    interval_seconds = 60

    rule {
        condition      = "B"
        for            = "0s"
        labels         = { }
        name           = "Foobar test"

        data {
          datasource_uid = "test"
          ref_id = "A"
          relative_time_range {
            from = 3600
            to = 0
          }
          model = "{}"
        }
    }
    
    rule {
        condition      = "B"
        for            = "0s"
        labels         = { }
        name           = "Foobar"

        data {
          datasource_uid = "test"
          ref_id = "A"
          relative_time_range {
            from = 3600
            to = 0
          }
          model = "{}"
        }
    }
}

```


Plan for the second run:
```
 # module.grafana_prod.grafana_rule_group.default will be updated in-place
  ~ resource "grafana_rule_group" "default" {
        id               = "7WfjWOKVk;default"
        name             = "default"
        # (3 unchanged attributes hidden)

      ~ rule {
          ~ name           = "Foobar" -> "Foobar test"
            # (7 unchanged attributes hidden)

          ~ data {
              ~ model          = jsonencode(
                  ~ {
                      - intervalMs    = 1000 -> null
                      - maxDataPoints = 43200 -> null
                    }
                )
                # (2 unchanged attributes hidden)

                # (1 unchanged block hidden)
            }
        }
      + rule {
          + condition      = "B"
          + exec_err_state = "Alerting"
          + for            = "0s"
          + name           = "Foobar"
          + no_data_state  = "NoData"

          + data {
              + datasource_uid = "test"
              + model          = jsonencode({})
              + ref_id         = "A"

              + relative_time_range {
                  + from = 3600
                  + to   = 0
                }
            }
        }
    }
```
**UPD:** fixed code for the first apply

**Comment by hilbert-ralf** [comment]
+1 as i facing the same issue. Is there any workaround i could apply until this got patched?

**Comment by tomersa** [comment]
I'm having the same issue. Hope it will be fixed soon 🙏

**Comment by tomersa** [comment]
> +1 as I facing the same issue. Is there any workaround I could apply until this got patched?

@hilbert-ralf  My workaround is to just destroy and apply again.

**Comment by hilbert-ralf** [comment]
> > +1 as I facing the same issue. Is there any workaround I could apply until this got patched?
> 
> @hilbert-ralf My workaround is just destroy and apply again.

thx @tomersa , should not be necessary but worked for me too.

**Comment by alexweav** [comment]
Found the root cause - It looks like a re-ordering issue in the Grafana API, in some logic that manages rule groups. What happens is the rules in your state are in a different order from what's specified in your file - TF tries to fix this ordering on apply (which is the right thing to do), but it hits this re-ordering issue, and the request is rejected.

This is also why I failed to reproduce it earlier - my local state had the rules in a different order. It sent the same request, but did not attempt to re-order the rules. I'll file an issue in the `grafana/grafana` repo.

**Comment by olekgo** [comment]
> +1 as i facing the same issue. Is there any workaround i could apply until this got patched?

My workaround is to put every rule into a separate group (arrived to the same conclusion as [here](https://github.com/grafana/terraform-provider-grafana/issues/711#issuecomment-1341436429)). GUI-wise doesn't look tidy, but otherwise seems to work well.

**Comment by billcchung** [comment]
We had this issue too, our workaround is by using `null_resource`  that keeps tracks of the number of alerts within a group, then use `replace_triggered_by` so it gets replaced whenever the number of alerts changed.:

```
resource "null_resource" "alerts" {
  triggers = {
    count = length(local.alerts)
  }
}
```

and we have alerts defined as a list:
```
local {
  alerts = [
    {
      name        = "alert1",
      metric_expr = "...",
    },
    {
      name        = "alert2",
      metric_expr = "...",
    },
  ]
}
```

and the group looks like:

```
resource "grafana_rule_group" "rule_group" {
  provider         = grafana.grafana
  org_id           = 1
  name             = "group_name"
  folder_uid       = grafana_folder.cloud.uid
  interval_seconds = 300

  dynamic "rule" {
    for_each = local.alerts
    content {
      name = rule.value.name
      ...other attributes...
    }
  }

  lifecycle {
    replace_triggered_by = [null_resource.alerts]
  }

}
```

This forces the group to be replaced everytime we add/remove alerts, and we have the `alerts` list because we dont want to write a bunch of same rules so we group similar alerts together to avoid duplicating the same contents, hope this can help someone with the same issue.

This is not without drawbacks, if you have alerts firing then those alerts will be resolved when they got removed.

**Comment by alexweav** [comment]
Here is the exact root cause of this in grafana/grafana: https://github.com/grafana/grafana/issues/66158

I spent some time trying to create a workaround to this in the provider today. Unfortunately, it only exposed more issues - I found at least 3 different ways this bug can manifest, and several of my workaround approaches had undesirable effects on existing plans. I think closing ^ is probably the fastest path forward to resolving this bug. Thanks for your patience everyone.

**Comment by Trojan295** [comment]
I have the same problem, I'm using `dynamic` block to create multiple alerts from a list.

```terraform
variable "metrics_alert_threshold" {
  type = list(object({
    name  = string
    value = number
  }))
  default = [
    {
      name  = "Over 100k"
      value = 100000
    },
    {
      name  = "Over 50k"
      value = 50000
    }
  ]
}


resource "grafana_rule_group" "metrics_alert" {
  name             = "Metrics usage"
  folder_uid       = grafana_folder.usage.uid
  interval_seconds = 240
  org_id           = 1

  dynamic "rule" {
    for_each = var.metrics_alert_threshold
    content {
      name           = rule.value.name
      for            = "2m"
      condition      = "A"
      no_data_state  = "NoData"
      exec_err_state = "Alerting"

      is_paused = false
      data {
        ref_id     = "A"
        query_type = ""
        relative_time_range {
          from = 600
          to   = 0
        }
        datasource_uid = data.grafana_data_source.prometheus.uid
        model = jsonencode({
          hide          = false
          intervalMs    = 1000
          maxDataPoints = 43200
          refId         = "A"
          expr          = "grafanacloud_org_metrics_billable_series > ${rule.value.value}"
        })
      }
    }
  }
}
```

I if add/remove any element in the `metrics_alert_threshold` list I get the error:
```
│ Error: status: 500, body: {"message":"failed to update alert rules: a conflicting alert rule is found: rule title under the same organisation and folder should be unique","traceID":"1ee0f5b19c40bc0e"}
│ 
│   with grafana_rule_group.metrics_alert,
│   on alerting.tf line 9, in resource "grafana_rule_group" "metrics_alert":
│    9: resource "grafana_rule_group" "metrics_alert" {
│ 
```

**Comment by JacobsonMT** [comment]
The root cause [issue](https://github.com/grafana/grafana/issues/66158) mentioned above has been fixed with https://github.com/grafana/grafana/pull/67868. It should be available in Grafana v10.1.X.

**Comment by julienduchesne** [comment]
Closing this now. Re-open or leave a comment if this is still an issue in Grafana v10.1+
